File: src/dataset/cpp/binary-search-tree_exercism-toemm/openAI/ChatGPT4_binary_search_tree_v3.cpp
#include "binary_search_tree.h"

namespace binary_search_tree {

using std::unique_ptr;

struct Node {
    int value;
    unique_ptr<Node> left;
    unique_ptr<Node> right;

    Node(int val) : value(val), left(nullptr), right(nullptr) {}
};

class BinarySearchTree {
private:
    unique_ptr<Node> root;

    void insertHelper(unique_ptr<Node>& node, int value) {
        if (!node) {
            node = std::make_unique<Node>(value);
            return;
        }
        if (value < node->value) {
            insertHelper(node->left, value);
        } else if (value > node->value) {
            insertHelper(node->right, value);
        }
    }

    bool searchHelper(const unique_ptr<Node>& node, int value) const {
        if (!node) return false;
        if (node->value == value) return true;
        return value < node->value ? searchHelper(node->left, value) : searchHelper(node->right, value);
    }

    void inOrderHelper(const unique_ptr<Node>& node, std::vector<int>& result) const {
        if (!node) return;
        inOrderHelper(node->left, result);
        result.push_back(node->value);
        inOrderHelper(node->right, result);
    }

public:
    BinarySearchTree() : root(nullptr) {}

    void insert(int value) {
        insertHelper(root, value);
    }

    bool search(int value) const {
        return searchHelper(root, value);
    }

    std::vector<int> inOrderTraversal() const {
        std::vector<int> result;
        inOrderHelper(root, result);
        return result;
    }
};

}  // namespace binary_search_tree

File: src/dataset/javascript/binary-search-tree_exercism-ThomasZumsteg/openAI/ChatGPT4_binary-search-tree_v3.js
class BinaryTree {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }

    insert(data) {
        const node = data <= this.data ? 'left' : 'right';
        if (this[node]) {
            this[node].insert(data);
        } else {
            this[node] = new BinaryTree(data);
        }
    }

    each(func) {
        if (this.left) this.left.each(func);
        func(this.data);
        if (this.right) this.right.each(func);
    }
}

module.exports = BinaryTree;